### PR TITLE
audit log postgres session PID

### LINKED ISSUE #38820: Audit log postgres connection PID
What would you like Teleport to do?
* Audit log the PID when a new postgres session is established.

What problem does this solve?
* Allows customers to associate a running process to a teleport user's database session. Since users may (and likely do) share a database user, the database user who started the query is not enough information. While PID is also not complete information, it can help to quickly narrow down the origin of a running postgres backend process.

If a workaround exists, please include it.
* N/A

### PR DESCRIPTION
**Pull request by GavinFrazar**
Changelog: Postgres database session start events now include the Postgres backend PID for the session.

Closes https://github.com/gravitational/teleport/issues/38820
- https://github.com/gravitational/teleport/issues/38820

### CODE REVIEW

File: api/proto/teleport/legacy/types/events/events.proto [hunk 1]
```diff
@@ -3041,6 +3041,9 @@ message DatabaseSessionStart {
     (gogoproto.embed) = true,
     (gogoproto.jsontag) = ""
   ];
+  // PostgresPID is the Postgres backend PID that was created for a Postgres
```
Comment (flyinghermit) [inline api/proto/teleport/legacy/types/events/events.proto:3044]: Can I ask for some more details on the `Postgres backend PID` especially, 
- The scope of the ID within Postgres
- What type of value it contains (`uuid` or just a random chars). Looking at `web/packages/teleport/src/Audit/fixtures/index.ts` file below, looks like its just a random string.
- Is it a sharable ID or is security sensitive?

File: web/packages/teleport/src/Audit/fixtures/index.ts [hunk 2]
```diff
@@ -1625,6 +1625,25 @@ export const events = [
...
+  {
+    cluster_name: 'root',
+    code: 'TDB00I',
+    db_name: 'test',
+    db_protocol: 'postgres',
+    db_service: 'local',
+    db_uri: 'localhost:5432',
+    db_user: 'postgres',
+    ei: 0,
+    postgres_pid: 123,
```
Comment (greedy52) [inline web/packages/teleport/src/Audit/fixtures/index.ts:1637]: nit: Is this shown anywhere on audit page? if not, i don't think it's very useful to define an example here.

**Review by greedy52 — APPROVED** [review]

File: web/packages/teleport/src/Audit/fixtures/index.ts (hunk 2 above)
Comment (GavinFrazar) [reply web/packages/teleport/src/Audit/fixtures/index.ts:1637]: oh, no it's not displayed on the overview/list of audit events. it's just show in an individual event.

File: api/proto/teleport/legacy/types/events/events.proto (hunk 1 above)
Comment (GavinFrazar) [reply api/proto/teleport/legacy/types/events/events.proto:3044]: > The scope of the ID within Postgres

I'm not sure what you mean by scope, but a brief summary of this ID: It's the PID of background process that the postgres server has started to handle the connection. Each PID can only handle one thing at a time, so for example a user might make a query like `SELECT pg_sleep(99999)` and they won't be able to do anything with that connection/PID until the statement completes (or they cancel it).

> What type of value it contains

It's always uint32

> Is it a sharable ID or is security sensitive?

It is not a sensitive value and can be shared. Postgres enforces controls around who can cancel/terminate a PID:
https://www.postgresql.org/docs/16/functions-admin.html#FUNCTIONS-ADMIN-SIGNAL
>> This is also allowed if the calling role is a member of the role whose backend is being canceled or the calling role has privileges of pg_signal_backend, however only superusers can cancel superuser backends.

There is a secret value exchanged at the start of the connection which also must be supplied to cancel a PID, but we do not log that value.

File: api/proto/teleport/legacy/types/events/events.proto [hunk 3]
```diff
@@ -3041,6 +3041,9 @@ message DatabaseSessionStart {
...
+  // connection.
+  uint32 PostgresPID = 8 [(gogoproto.jsontag) = "postgres_pid,omitempty"];
```
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply api/proto/teleport/legacy/types/events/events.proto:3046]: > curious if there is anything else we think we might capture in the future for PostgreSQL

Nothing comes to mind

File: lib/srv/db/postgres/engine.go [hunk 4]
```diff
@@ -147,6 +147,7 @@ func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Sessio
 		cancelAutoUserLease()
 		return trace.Wrap(err)
 	}
+	sessionCtx.PostgresPID = hijackedConn.PID
```
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/srv/db/postgres/engine.go:150]: Yeah that's a good idea, like this right?

```go
	sessionCtx.PostgresPID = hijackedConn.PID
	e.Log = e.Log.With("pg_backend_pid", hijackedConn.PID)
```

File: web/packages/teleport/src/Audit/fixtures/index.ts (hunk 2 above)
Comment (GavinFrazar) [reply web/packages/teleport/src/Audit/fixtures/index.ts:1637]: Removing this, I see what you mean

File: lib/srv/db/postgres/engine.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/srv/db/postgres/engine.go:150]: added in https://github.com/gravitational/teleport/compare/0d8c35fd629323320b333465e8563a479fb21d6a..4d29cc809e60f6ebcd5ff912051b6f5a2e24afcc

File: api/proto/teleport/legacy/types/events/events.proto (hunk 1 above)
Comment (flyinghermit) [reply api/proto/teleport/legacy/types/events/events.proto:3044]: Got it. Sorry for not being clear on "The scope of the ID within Postgres". I meant to ask where would the `ID` be used within the Postgres and for what purpose. I think you answered that with summary. 

I would add those description in the comment too. The reason I asked those was previously we logged some sensitive session IDs in the log. This one looks good.

**Review by flyinghermit — APPROVED** [review]

File: api/proto/teleport/legacy/types/events/events.proto (hunk 1 above)
Comment (GavinFrazar) [reply api/proto/teleport/legacy/types/events/events.proto:3044]: I added to the comment for the protobuf

**Review by ravicious — APPROVED** [review]

**Pull request merged**